src/main/java/bsh/Operators.java

```diff
@@ -515,26 +570,28 @@ static Object[] promotePrimitives(Object lhs, Object rhs)
         Number lnum = promoteToInteger(lhs);
         Number rnum = promoteToInteger(rhs);
 
-        if ( lhs instanceof BigDecimal ) {
-            if ( !(rhs instanceof BigDecimal) )
-                rhs = Primitive.castNumber(BigDecimal.class, rnum);
-        } else if ( rhs instanceof BigDecimal ) {
-            lhs = Primitive.castNumber(BigDecimal.class, lnum);
-        } else if ( Types.isFloatingpoint(lhs) || Types.isFloatingpoint(rhs)) {
-            if ( !(lhs instanceof Double) )
-                lhs = Double.valueOf(lnum.doubleValue());
-            if ( !(rhs instanceof Double) )
-                rhs = Double.valueOf(rnum.doubleValue());
-        } else if ( lhs instanceof BigInteger ) {
-            if ( !(rhs instanceof BigInteger) )
-                rhs = Primitive.castNumber(BigInteger.class, rnum);
-        } else if ( rhs instanceof BigInteger ) {
-            lhs = Primitive.castNumber(BigInteger.class, lnum);
-        } else {
-            if ( !(lhs instanceof Long) )
-                lhs = Long.valueOf(lnum.longValue());
-            if ( !(rhs instanceof Long) )
-                rhs = Long.valueOf(rnum.longValue());
+        if (!(lhs instanceof Float && rhs instanceof Float)) {
```

Comment:
It is hard to see what happens here but this seems wrong.

From the existing function:
https://github.com/beanshell/beanshell/blob/53b0436f4e2777e5ec8c8f60e565c7ef48f3b18b/src/main/java/bsh/Operators.java#L513-L541

There is a precedence BigDecimal, Double, BigInteger, Long and it seems this check needs to be  around line 524 when we are checking for Doubles.

Reply:
I can refactor this but the outcome will be the same.